### ISSUE TITLE
Consider displaying configuration dialog fields without using a table

### ISSUE THREAD

**Issue by mmcev106** [vanderbilt/redcap-external-modules#184]
The current table design causes field label and input widths to all interact with each other.  This means if you have one long label (like a paragraph), ALL inputs become smaller.  Ideally only the input for that row would be smaller.

**Comment by mmcev106** [comment]
Before anyone works on this, we should reach out to @grezniczek.  He was considering a redesign of the configuration dialog.

**Comment by grezniczek** [comment]
Thanks for reminding me.
@mmcev106  and I had a long email conversation last October about re-doing the module configuration UI. I had brought it up because of a rather complex configuration of a module I was working on (still not finished) that cannot be handled by the current implementation, as for one, it would be a usability nightmare (lots of nested and repeated blocks) and there were (still are?) bugs in the branching logic when nesting is involved.

### Here is a brief summary of the main points from our discussion.

**A new configuration management dialog should (items in no particular order):**
- be user friendliy and look good.
- support anything that is currently described in the config.json documentation, including arbitrarily deep nesting with branching logic.
- ideally be fully backward compatible with all released modules (this may not be possible in some cases for modules employing custom Javascript to build custom fields or modify the current dialog).
- allow for the future possibility of modules defining completely custom fields while staying compatible with core features (repeatables, subsettings, etc.).
- provide mechanisms for admins and project designers to easily export/backup or import settings.
- should be capable and sexy enough so module authors don't have to come up with their own config implementations.

### Before anything get's coded, however, we would like to
- have an open design phase with the community,
- collect feature requests, so that even when not considered initially, the dialog can be designed in a way that is open to there implementation later,
- provide opportunity for feedback from the community

### To get things started, here are some ideas
- For the possible look and feel, consider [this design study](https://github.com/grezniczek/redcap_em_config_study). It's implemented as a REDCap EM that provides the (more or less static) mock-up as a plugin page in Control Center. The README has some screenshots, and a walkthrough, _so definitely do check it out!_
- The new config dialog is essentially a drop-in replacement for the old dialog (also from the perspecitve of the EM-framework - data would be submitted for storage as today, so no backend changes planned). Everything happens in the browser (with the help of a few ajax requests back to the server when/if necessary).
- Regarding the import/export of settings, there have been [some discussions on the community site](https://community.projectredcap.org/idea/72500/new-module-under-construction-move-module-config.html). The new dialog should support export and best-effort import of the complete settings or starting at a certain level in a nested settings hierarchy (e.g. only the values of one particular repeating block). As mentioned in that community thread, this should support importing settings saved from a different project, which means there must be some (user-guided) mapping to fit the new environment (project ids, events, arms, etc.).
- The syntax for declaring config options in config.json should probably be extended to support things like tabs (as shown in the mockup), and maybe to declare "private" fields (that are set by modules automatically but might be relevant for backup), or sensitive fields (e.g. containing credentials for othe systems) that then maybe cannot be exported (and neither be imported) (such processing would always occur on the server).
- For custom fields, the dialog would provide a "canvas" for the a module to "draw" on, and a backing field for the data to be stored, as well as probably a JS-API a custom field needs to implement in order to support import/export.

_Let's have some discussion!_

**Comment by 123andy** [comment]
Some ideas:

1) Some sort of module-wide JS object that supports some common hooks.  An example is when a config page is rendered you could make an ajax call to your module to determine more detailed debugging and configuration options...  When an EM is complex, it is useful to be able to display more informative config errors or notices.

As an example, we also have a module that in the config it lets you add instruments to your project (required by the module) with a single click.  This is kind of handy and is currently hacked into the config modal...

2) I like the "i" icon - maybe this is a new field property that includes more verbose details about the configuration of a field...

3) A common motif is "Project -> Event -> Field" when specifying an external project.  It would be nice to have this in a single 'field' type that allows for quick navigation in one field...

Great work

**Comment by alanisem123** [comment]
- If the goal is to make the config-modal better looking, then I think you got it.
- If the goal is to make the config-modal more functional, then that's a different question. 
- I think the underlying configuration, of the proposed solution, should definitely be kept in a json file.
- I would build on top the existing config.json keys to determine what else is needed.

**Comment by 123andy** [comment]
FWIW - I know Ihab on my team fixed the branching logic issue with the config but that is probably getting redone anyhow...  I think the branching logic is the biggest UI confounder right now with real-time 'valid config' updates.  Maybe there could be an ajax callback as an optional parameter that returns an array that could render a green alert saying config valid or some user-specified message or a red box with bullet points of what needs to get fixed for the config to be valid...

**Comment by grezniczek** [comment]
@alanisem123, the goal is both. To make the config dialog more user-friendly in case of complex configurations (this will include "better looking"), and to make it more functional.

Regarding the latter, this would include:
- adding mechanisms to import/export settings (and thus facilitating transfer of settings between projects and/or REDCap instances).
- adding validation (and messages), as @123andy suggested.
- maybe adding new field types (like for project/event/field as @123andy suggested).

Regarding implementation: I guess JSON would be the obvious candidate for import/export as well as for any communication between the modal and the backend. 

Existing config.json stay intact, and the new dialog would fully support any existing module, but new keywords would be added (like for the translation support recently). Candidate additions in the scope of config items would be _description_ (rendered as short explanatory text below whatever is in _name_), _help_ (additional info that is rendered as a popup for the i-icon), and _tab_ (for grouping related options into tabs).

Regarding support for better validation, the dialog (and respective additions to the EM framework) would support better validation, probably via a hook similar to the existing "redcap_module_save_configuration" hook, maybe _redcap_module_validate_configuration_, which is called called before values are saved and which gets passed a complete configuration that can be annotated by the module (thus being able to give detailed error information). The module would indicate via config.json whether it supports this validation (maybe by including the hook name in the "permissions" section).

**Comment by mmcev106** [comment]
@taylorr4 and I just spoke about this briefly and wanted to stress that a drastic change to the configuration interface may be difficult to approve in one pass.  It may be beneficial to make changes in phases to ease the testing/review process, and the adjustment process for users.

**Comment by grezniczek** [comment]
My idea would have been to implement this in a fully backward compatible way (regarding config.json) and to offer the new config dialog in parallel to the current dialog.

This could be realized with a second button or with a toggle on the EM manager pages that switches between them. So this could mature (in testing; and maybe even in production for a while) before replacing the current dialog. 

I am currently working of a module-specific config (a plugin page) that uses some of the features envisioned for the new config dialog and gave me a few insights. Not sure though when I will find time to flesh out some logic for the "EM Config Study".

**Comment by dr01d3r** [comment]
An `Enhanced Configuration` button (in addition to the existing config button) can be a good thing, so long as there is commitment to transition.  A vendor LIMS I support implemented a shiny new "Enhanced Explorer" for everyone to transition to, while preserving the original explorer feature.  It's been nearly 10 years and both buttons are still there.

I'm in favor of the new interface, but I also get Vandy's position on change.  When you basically have the whole thing done, how would we go about piecing it apart?

**Comment by grezniczek** [comment]
What do you mean with "piecing it apart", @dr01d3r? Taking individual features from it and adding them piece by piece to the old one?

I don't think we would or should do that. If some new field types the old system cannot handle would be introduced with the new `Enhanced Configuration` (I like the name), they might just not show up in the old config and if people want to use a module that needs those, they'll just have to "put up" with the new config (and hopefully never go back afterwards ;) 
But at first, it will probably be mostly a cosmetic change only (tabs and more help text being the main features).

**Comment by 123andy** [comment]
Thanks for your time and effort, @grezniczek .    I think a 'dual' button with a cookie would be a great start.  Once we have confirmed that things are relatively bug-free we can deprecate the old version.

I know, for one, that some of our modules that hacked on javascript to improve the interface will have to be re-done...

I also like the idea of a 'config page' option in config.json where you could do more custom configs for those modules where the default is just not enough to handle the complexity...

**Comment by pbchase** [comment]
@123andy, rather than cookie-based config, this feels more like a module-level configuration to me. If the EM framework supported both the old and the new config methods, module writers could choose their configuration method. I envision a new "configuration version" much like the framework version. I can think of a few of my modules that would benefit from a redesigned interface. I would cheerfully modify them to make them easier to config. 

Meanwhile, older modules that did not specify "configuration version" would run the older config code. At some point the EM Framework team might decide the were done supporting the original config code. They could remove it in a new framework release to mark this breaking change.

**Comment by dr01d3r** [comment]
@grezniczek I just meant if we were to try and ease into a full implementation.  I dont know how that would happen.  And to my point before about how variable upgrade cycles are, it might not be any different than a full release.

I like the version concept or letting new modules adopt the new config.  With this though, we're committing to another version min that will restrict our ability to stay backward compatible.  If it is just embedded into REDCap and modules are ignorant of it, it gives a little more flexibility that way.

**Comment by 123andy** [comment]
I like it, Phillip.  As for backward compatible if you're using EMs, you should be committed to staying relatively current so I'm not so worried about encouraging some to upgrade more often...

**Comment by ChemiKyle** [comment]
On the topic of import/export of module config I would be happy to help in any way I can. The logic from my [Move Module Config](https://github.com/ChemiKyle/move_module_config/tree/develop) module - though fairly unpolished - should integrate well into this.
